Add NucleusX Model

syncdoth opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

This PR adds a new model named NucleusX. This model is contributed by [Sehyun Choi](https://github.com/syncdoth) and [NucleusAI](https://www.withnucleus.ai/). The model is based on the [Retentive Network](https://arxiv.org/abs/2307.08621) architecture, and the code is largely adapted from [this repo](https://github.com/syncdoth/retnet.git), which again borrows core implementations from [torchscale](https://github.com/microsoft/torchscale). We are planning to release our paper and weights soon.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

We kindly request the review of this new model from @ArthurZucker and @younesbelkada!

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

Commit:
[WIP] Add NucleusX Model

Commit:
solve chunkwise equivalence and add test

Commit:
ignore attention checking for nucleusx

Commit:
update license

Commit:
implement generate from input_embeds

Commit:
handle weight initialization issue

Commit:
update NucleusX docs

Commit:
fix typo; init rmsnorm in _init_weights

Commit:
refactor relpos, check use_cache with gradient_checkpoint, default activation_fn=swish

Commit:
necessary changes for code styling

Commit:
fix nucleusai url

Commit:
remove nucleus_x._set_gradient_checkpointing

Commit:
remove timm dependency and copy drop_path

Commit:
resolve unused config, init parameter of NucleusXModels

syncdoth commented on src/transformers/models/nucleus_x/configuration_nucleus_x.py:
```diff
@@ -0,0 +1,211 @@
+    This is the configuration class to store the configuration of a [`~NucleusXModel`]. It is used to instantiate an
+    NucleusX model according to the specified arguments, defining the model architecture. Instantiating a configuration
+    with the defaults will yield a similar configuration to that of the NucleusX-7B
+    [NucleusAI/NucleusX-7B](https://huggingface.co/NucleusAI/NucleusX-7B) architecture.
```
This weight is not released yet; we are planning to release the weight at this link soon! We have included this link to pass the configuration testing requiring a link to the checkpoint.

Commit:
add reset_parameters to NucleusXRMSNorm

Commit:
refactor past_key_value to tuple, update docs and tests accordingly

Commit:
refactor argument docs, adds to _toctree

Commit:
add empty returns

Commit:
put nucleus_x under text model in _toctree

Commit:
update documents

syncdoth commented:
The current test failure at `tests_pr_documentation_tests` is due to the incorrect repo_id and links, namely `NucleusAI/NucleusX-7B` and https://huggingface.co/NucleusAI/NucleusX-7B used in examples and model docs. These checkpoints are not released yet; we plan to release them soon.

syncdoth commented:
cc: @sippycoder and also @LysandreJik!

Commit:
revert added comma

Removes the accidentally added comma in tests/generation/test_utils.py

ArthurZucker commented:
Hey! Thanks for opening the PR, I'll let @Rocketknight1 do a first review as he is more familiar with this kind of models!

Commit:
better handling of dtype for half precision weights

When loading the model weights in different dtype other than fp32,
`.float` statements may cause troubles. This commit handles the tensor
creations and float casting to be aware of the `dtype` of the weights.

Rocketknight1 commented:
Hi all! RetNets seem like a really interesting architecture, so I'm quite excited to take a look - I'll try to review this in the next day or two.

Commit:
rename layernorm -> rms_norm

Commit:
remove subln

Reason: since we are using GLU, sub-layernorm is not well-defined.

Commit:
test parallel vs recurrent with 10 random examples

Rocketknight1 commented on src/transformers/models/nucleus_x/modeling_nucleus_x.py:
```diff
@@ -0,0 +1,1449 @@
+        else:
+            raise ValueError(f"forward_mode {forward_mode} not supported.")
+
+        # concaat heads
```
```suggestion
        # concat heads
```

Rocketknight1 commented on src/transformers/models/nucleus_x/modeling_nucleus_x.py:
```diff
@@ -0,0 +1,1449 @@
+        else:
+            self.alpha = 1.0
+
+    def build_ffn(self):
+        return NucleusXGLU(
+            self.embed_dim,
+            self.ffn_dim,
+            self.config.activation_fn,
+            self.config.dropout,
+            self.config.activation_dropout,
+        )
```
I have a very slight preference to remove this method and just inline it in the init: `self.ffn = NucleusXGLU(...)`

Rocketknight1 reviewed (commented):
Hi all, I just looked through this! Overall, the core modelling code looks very solid, and I couldn't find much to complain about. We normally encourage the use of `# Copied from` in these PRs, but given that RetNets differ significantly from Transformers, most functions here will be unique to NucleusX.

I also think the test coverage is good, in particular the tests confirming that outputs are equivalent in parallel/recurrent/chunkwise mode.

Before we can merge, though, we need checkpoints to be uploaded. Also, we ideally need an integration test. The purpose of these tests is to ensure that model output for a specific checkpoint remains numerically constant, which is very important to ensure that future updates don't create silent errors. [Here is an example](https://github.com/huggingface/transformers/blob/main/tests/models/deberta_v2/test_modeling_deberta_v2.py#L309) of an integration test that you can copy for NucleusX. 
If your checkpoints are too large for our CI, we can make a `tiny-random-nucleusx` model to use for the integration test. An integration test confirming generation output remains constant when `do_sample=False` would also be helpful!

Overall though, this looks like a really solid PR, and I suspect we shouldn't have much trouble including this in `transformers`. Thank you for your contribution!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_27259). All of your documentation changes will be reflected on that endpoint.

Rocketknight1 commented on src/transformers/models/nucleus_x/configuration_nucleus_x.py:
```diff
@@ -0,0 +1,211 @@
+    >>> from transformers import NucleusXModel, NucleusXConfig
+
+    >>> # Initializing a NucleusX-7B style configuration
+    >>> configuration = NucleusXConfig()
```
```suggestion
    >>> configuration = NucleusXConfig(decoder_layers=2)
```
Also, one more comment! The doctest runner is crashing on this file, and I suspect the reason is that it's running out of memory because you're initializing a 7B model in float32 and so using 28GB of memory, which is a lot for the doctest runner! Maybe change this line to initialize a much smaller model?

syncdoth replied:
I was trying to find out why the doctest is failing for a long time, and this makes total sense!

Commit:
remove build_rms_norm, redefine NucleusXRMSNorm

This follows the example of other models, such as LongT5, idefics, llama, etc.

Commit:
fix typo, inline init of ffn, smaller example model size

This resolves the comments by @Rocketknight1

syncdoth commented:
@Rocketknight1 Thanks for reviewing this PR! I have gone through the comments and resolved them. There are also some other updates:

- `dtype` handling for tensors created in `NucleusXRelPos` to have the same `dtype` as the model weights
- rename some `*layer_norm` modules to `*rms_norm` for conformity.
- removed `subln` option (Sub-LayerNorm), which is not applicable to our choice of FFN (SwiGLU).

There are other minor changes, which can be found in the commit logs. Thanks again!

syncdoth commented on tests/models/nucleus_x/test_modeling_nucleus_x.py:
```diff
@@ -131,6 +131,7 @@ def get_config(self):
             decoder_layers=self.num_hidden_layers,
             is_decoder=False,
             decoder_retention_heads=self.num_attention_heads,
+            use_cache=False,
```
When `use_cache=True`, `NucleusXMultiScaleRetention.parallel_forward` will be less efficient. This is because `use_cache=True` makes the `NucleusXMultiScaleRetention.parallel_forward` to compute `past_key_values`, which incurs another `O(T^2)` computations.

Only when we want to do recurrent forward following the parallel forward (e.g. during generation, we compute the prompt in parallel, but generate with recurrent mode), `use_cache=True` should be set.

Rocketknight1 replied:
cc @gante to this bit - is our generation code ready to handle networks with multiple forward modes?

Rocketknight1 commented:
Also @syncdoth, while we're waiting for the checkpoints, there are some tests failing in this PR that are unrelated. If you pull the latest version from main and rebase your PR, that should fix them.

syncdoth commented on src/transformers/models/nucleus_x/modeling_nucleus_x.py:
```diff
@@ -0,0 +1,1441 @@
+
+        forward_mode = kwargs.get("forward_mode", "parallel")
+        if past_key_values is not None:
+            # NOTE: when we have past_key_values, using recurrent mode will be faster.
```
@Rocketknight1 @gante For your question about "generate code handling networks with multiple forward mode" in another comment, this is my take on that: when we call `prepare_inputs_for_generation`, if we detect `past_key_values`, it means that the prompt to the generation has been computed, and using recurrent forward is better. Hence the `forward_mode = "recurrent"` line below.

Note that `forward_mode` is just a string that the model takes as forward input to select the forward mode at each forward step!

gante replied:
@Rocketknight1 @syncdoth We can support multiple generation models, but the implementation for it depends on a few factors! The audio models are the best examples. For instance:
1. Whisper wraps `generate` and accepts additional flags. These flags trigger additional arguments for `generate` ([e.g. a custom logits processor to generate timestamps](https://github.com/huggingface/transformers/blob/b86c54d9ff4504b7287d95a215f8d6fa9388761f/src/transformers/models/whisper/modeling_whisper.py#L1965)) or postprocessing
2. Bark contains 3 internal models, and wraps `generate` to call `generate` on them in a sequence

Additionally, if `model.forward` accepts multiple modes, you can also prepare the flags in `model. prepare_inputs_for_generation`, as written above :)

syncdoth commented:
> Also @syncdoth, while we're waiting for the checkpoints, there are some tests failing in this PR that are unrelated. If you pull the latest version from main and rebase your PR, that should fix them.

This may be a beginner question, but should I rebase main and (force?) push or merge main and push?

Rocketknight1 commented:
Probably the easiest way to do it is to pull the latest version of main, then rebase your branch onto main, and then force push.

fakerybakery commented:
Hi @syncdoth, do you know what happened to Nucleus AI? The website is now down

syncdoth commented:
> Hi @syncdoth, do you know what happened to Nucleus AI? The website is now down

This is unrelated to this PR but there's some maintenance going on with the website. Hang tight :)

Rocketknight1 commented:
btw @syncdoth if you're still getting test failures, try 'sync upstream' on the `main` branch of your forked repo, then on your development machine, pull the latest main branch, change to the `add_nucleus_x` branch, rebase and finally force push. Should resolve everything!

Rocketknight1 commented:
Don't stale, please! This looks quite close to being ready! (cc @syncdoth - let me know if you need any help with the last bit!)

syncdoth commented:
We are on the verge of releasing the weight! There's been a bit of delay in the schedule 🥲

The last bit should be updating the weight links in the docs and writing the integration tests; We are working on it hard!

Commit:
update docs and checkpoint ref

Commit:
integration testing

syncdoth commented:
Hi @Rocketknight1, I’m seeing test failure related to document building, and testing the run of NucleusXForCausalLM.forward example. It seems that it might be due to `.from_pretrained` from a 7B checkpoint killing the worker, like the previous example in the configuration. Do you think I should change the example to a smaller one?

Maykeye commented:
Does it require some tinkering to use `generate` in not parallel mode?  (I don't have RAM for processing 16KB prompt in parallel)

I dumped source to model folder, edited config to treat it as `trusted_remoted_code=True` thingy, parallel works fine, as in [test](https://github.com/huggingface/transformers/blob/b624d0518776d1761d29a448b83de4b2ef4d6658/tests/models/nucleus_x/test_modeling_nucleus_x.py#L465):

```python
In [7]: print(tokenizer.decode(model.generate(**tokenizer("Hello my name is", return_tensors="pt").to("cuda"), max_new_tokens=20, do_sample=False, forward_mod
   ...: e="parallel").ravel()))
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.
/home/fella/src/llama/text-generation-webui/models/NucleusAI_Nucleus-X/modeling_nucleus_x.py:370: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  cache = (current_kv, scale, torch.tensor(prev_seqlen + 1, dtype=torch.long))
<s> Hello my name is Tina and I am a 25 year old female. I am a very outgoing person
```

but  recurrent no

```python
In [8]: print(tokenizer.decode(model.generate(**tokenizer("Hello my name is", return_tensors="pt").to("cuda"), max_new_tokens=20, do_sample=False, forward_mod
   ...: e="recurrent").ravel()))
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.
/home/fella/src/llama/text-generation-webui/models/NucleusAI_Nucleus-X/modeling_nucleus_x.py:370: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  cache = (current_kv, scale, torch.tensor(prev_seqlen + 1, dtype=torch.long))
<s> Hello my name is the most of the world.
The first thing I noticed was the size of the room. It
```

(Even if I say in config.json to use recurrent forward mode, 16KB prompt fails to pass through model.generate unless I use forward_mode='recurrent')

Rocketknight1 commented:
Hi @syncdoth, sorry for the Christmas delay! You're correct, though - the issue is almost certainly caused by the docstring trying to load a model too big for the test runner. Is there any smaller checkpoint we can use? You could also try `torch_dtype=torch.bfloat16`.

ArthurZucker commented:
No worries 🤗
